modules/xfem/src/userobjects/MeshCut2DUserObjectBase.C

```diff
@@ -0,0 +1,334 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "MeshCut2DUserObjectBase.h"
+
+#include "XFEMFuncs.h"
+#include "MooseError.h"
+#include "MooseMesh.h"
+#include "libmesh/edge_edge2.h"
+#include "libmesh/serial_mesh.h"
+#include "libmesh/mesh_tools.h"
+
+InputParameters
+MeshCut2DUserObjectBase::validParams()
+{
+  InputParameters params = GeometricCutUserObject::validParams();
+  params.addRequiredParam<MeshFileName>(
+      "mesh_file",
+      "Mesh file for the XFEM geometric cut; currently only the Exodus type is supported");
+  params.addClassDescription("Creates a UserObject base class for a mesh cutter in 2D problems");
+  return params;
+}
+
+MeshCut2DUserObjectBase::MeshCut2DUserObjectBase(const InputParameters & parameters)
+  : GeometricCutUserObject(parameters), _mesh(_subproblem.mesh())
+{
+  // only the Exodus type is currently supported
+  MeshFileName cutterMeshFileName = getParam<MeshFileName>("mesh_file");
+  _cutter_mesh = std::make_unique<ReplicatedMesh>(_communicator);
+  _cutter_mesh->read(cutterMeshFileName);
+  // test element type; only line elements are allowed
+  for (const auto & cut_elem : _cutter_mesh->element_ptr_range())
+  {
+    if (cut_elem->n_nodes() != 2)
+      mooseError("The input cut mesh should include EDGE2 elements only!");
+    if (cut_elem->dim() != 1)
+      mooseError("The input cut mesh should have 2D elements only!");
+  }
+
+  // find node fronts of the original cutmesh.  This is used to order everything
+  findOriginalCrackFrontNodes();
+}
+
+bool
+MeshCut2DUserObjectBase::cutElementByGeometry(const Elem * elem,
+                                              std::vector<Xfem::CutEdge> & cut_edges,
+                                              std::vector<Xfem::CutNode> & cut_nodes) const
+{
+  // With the crack defined by a line, this method cuts a 2D elements by a line
+  // Fixme lynn Copy and paste from InterfaceMeshCut2DUserObject::cutElementByGeometry
```

Comment:
I don't know.  It is a copy and paste and if we refactor one instance of it, we should refactor this as well.  Maybe they can be combined or it can be moved to a base class.  I don't think it is using any member data so maybe it can be a static.